frame/elections-phragmen/src/lib.rs

```diff
@@ -289,210 +434,158 @@ decl_error! {
+		ElectionError,
+		/// A \[member\] has been removed. This should always be followed by either `NewTerm` or
+		/// `EmptyTerm`.
+		MemberKicked(AccountId),
+		/// A \[member\] has renounced their candidacy.
+		MemberRenounced(AccountId),
+		/// A \[candidate\] was slashed by \[amount\] due to failing to obtain a seat as member or
+		/// runner-up.
+		///
+		/// Note that old members and runners-up are also candidates.
+		CandidateSlashed(AccountId, Balance),
+		/// A \[seat holder\] was slashed by \[amount\] by being forcefully removed from the set.
+		SeatHolderSlashed(AccountId, Balance),
+	}
+);
+
 decl_module! {
 	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
 		type Error = Error<T>;
-
 		fn deposit_event() = default;
 
 		const CandidacyBond: BalanceOf<T> = T::CandidacyBond::get();
-		const VotingBond: BalanceOf<T> = T::VotingBond::get();
+		const VotingBondBase: BalanceOf<T> = T::VotingBondBase::get();
+		const VotingBondFactor: BalanceOf<T> = T::VotingBondFactor::get();
 		const DesiredMembers: u32 = T::DesiredMembers::get();
 		const DesiredRunnersUp: u32 = T::DesiredRunnersUp::get();
 		const TermDuration: T::BlockNumber = T::TermDuration::get();
-		const ModuleId: LockIdentifier  = T::ModuleId::get();
+		const ModuleId: LockIdentifier = T::ModuleId::get();
 
 		/// Vote for a set of candidates for the upcoming round of election. This can be called to
 		/// set the initial votes, or update already existing votes.
 		///
-		/// Upon initial voting, `value` units of `who`'s balance is locked and a bond amount is
-		/// reserved.
+		/// Upon initial voting, `value` units of `who`'s balance is locked and a deposit amount is
+		/// reserved. The deposit is based on the number of votes and can be updated over time.
 		///
 		/// The `votes` should:
 		///   - not be empty.
 		///   - be less than the number of possible candidates. Note that all current members and
 		///     runners-up are also automatically candidates for the next round.
 		///
-		/// It is the responsibility of the caller to not place all of their balance into the lock
-		/// and keep some for further transactions.
+		/// The dispatch origin of this call must be signed.
+		///
+		/// ### Warning
+		///
+		/// It is the responsibility of the caller to **NOT** place all of their balance into the
+		/// lock and keep some for further transactions.
 		///
 		/// # <weight>
-		/// Base weight: 47.93 µs
-		/// State reads:
-		/// 	- Candidates.len() + Members.len() + RunnersUp.len()
-		/// 	- Voting (is_voter)
-		/// 	- Lock
-		/// 	- [AccountBalance(who) (unreserve + total_balance)]
-		/// State writes:
-		/// 	- Voting
-		/// 	- Lock
-		/// 	- [AccountBalance(who) (unreserve -- only when creating a new voter)]
+		/// we consider the common case of placing more votes. In other two case, we refund.
 		/// # </weight>
-		#[weight = T::WeightInfo::vote(votes.len() as u32)]
+		#[weight = T::WeightInfo::vote_more(votes.len() as u32)]
```

Comment:
It's my opinion that this should be `vote_more.max(vote_less).max(vote_equal)`

It is basically zero cost, and provides more customization and doesnt make assumptions about the weight of these functions.

Reply:
yeap I agree w Shawn, important to answer questions not create them.

Reply:
Not sure if I follow: `vote_more()` and `vote_less()`have additional state access (and arguably the same but that's a different story). If someone submits the same amount of votes, why should they pay for more expensive variant? doing a `.max()` is a backtrack to a more primitive version of weighing in which we just assume we always execute the most expensive path if the function. 

Even if we are to change, then it can stay as `#[weight = T::WeightInfo::vote_more(votes.len() as u32)]` without any of the refunds. `vote_more` is basically the max of all of them.

Reply:
> vote_more is basically the max of all of them.

I get that, but this is ultimately an assumption and something that may be configurable from the outside. Your initial weight here should always be the max of all possible paths, and you should use a simple max comparison to make that determination imo

Reply:
I will do as you recommend for now but I am not fully convinced yet.